daisy/disk.go

```diff
@@ -15,39 +15,127 @@
 package daisy
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"regexp"
+	"strconv"
 	"sync"
 
-	"github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/compute/v1"
 	"google.golang.org/api/googleapi"
 )
 
 var (
-	disks      = map[*Workflow]*diskRegistry{}
-	disksMu    sync.Mutex
+	diskCache struct {
+		exists map[string]map[string][]string
+		mu     sync.Mutex
+	}
 	diskURLRgx = regexp.MustCompile(fmt.Sprintf(`^(projects/(?P<project>%[1]s)/)?zones/(?P<zone>%[2]s)/disks/(?P<disk>%[2]s)$`, projectRgxStr, rfc1035))
 )
 
-type diskRegistry struct {
-	baseResourceRegistry
-	attachments      map[*resource]map[*resource]*diskAttachment // map (disk, instance) -> attachment
-	testDetachHelper func(d, i *resource, s *Step) dErr
+// diskExists should only be used during validation for existing GCE disks
+// and should not be relied or populated for daisy created resources.
+func diskExists(client daisyCompute.Client, project, zone, disk string) (bool, dErr) {
+	diskCache.mu.Lock()
+	defer diskCache.mu.Unlock()
+	if diskCache.exists == nil {
+		diskCache.exists = map[string]map[string][]string{}
+	}
+	if _, ok := diskCache.exists[project]; !ok {
+		diskCache.exists[project] = map[string][]string{}
+	}
+	if _, ok := diskCache.exists[project][zone]; !ok {
+		dl, err := client.ListDisks(project, zone)
+		if err != nil {
+			return false, errf("error listing disks for project %q: %v", project, err)
+		}
+		var disks []string
+		for _, d := range dl {
+			disks = append(disks, d.Name)
+		}
+		diskCache.exists[project][zone] = disks
+	}
+	return strIn(disk, diskCache.exists[project][zone]), nil
+}
+
+// Disk is used to create a GCE disk in a project.
+type Disk struct {
+	compute.Disk
+	resource
+
+	// Size of this disk.
+	SizeGb string `json:"sizeGb,omitempty"`
+}
+
+// MarshalJSON is a hacky workaround to prevent Disk from using compute.Disk's implementation.
+func (d *Disk) MarshalJSON() ([]byte, error) {
+	return json.Marshal(*d)
+}
+
+func (d *Disk) populate(ctx context.Context, s *Step) dErr {
+	errs := d.resource.populateWithZone(ctx, s, &d.Name, &d.Zone)
```

Comment:
d.Name, d.Zone, errs = d.resource.populateWithZone(ctx, s, d.Name, d.Zone)

or don't return name at all and instead:
d.Name = d.resource.realName

Reply:
Went with the first.